### ISSUE TITLE
Task assignee issue

### ISSUE THREAD

**Issue by jordotech** [codingjoe/joeflow#19]
I am currently adding assignees to a Task object via a `post_save` signal by matching the `task.name` against a string.  Is it possible to add an assignee from within a workflow method?

**Comment by jordotech** [comment]
I agree with you on point 1, I would opt to **not** using the request object there.  For what it's worth, the following is working quite nicely for me and does not introduce breaking changes.  

### Model

I made a new model where you configure users to task names:

```
class TaskAssignee(TimeStampedModel):
    role = models.CharField(
        "Role Name",
        max_length=50,
        help_text="Ex. 'Final Approver', 'Preliminary Approver'",
    )
    task_id = models.CharField(
        "Task ID",
        max_length=50,
        help_text="The machine name of the human workflow task that will be assigned to this approver.",
    )
    content_type = models.ForeignKey(
        "contenttypes.ContentType",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        limit_choices_to=workflow_state_subclasses,
        related_name="joeflow_assignee_set",
    )
   # todo: Add FK to content type
    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE,)

    class Meta:
        verbose_name = _("Task Assignee")
        verbose_name_plural = _("Task Assignees")

    def __str__(self):
        return f"{self.role}"
```
The problem with the above is that I'm manually entering the task method name into the admin form which is not very elegant.  Maybe the field could be a choices field that pulls options from a new setting

```
JOEFLOW_ASSIGNABLE_TASKS = ["prelim_approve", "final_approve"]
```

### Signal

Then a `post_save` signal that adds the assignees at Task creation if there's a match.

```
@receiver(post_save, sender=Task)
def assign_task(sender, instance, created, raw, using, update_fields, **kwargs):
        if created:
        if instance.type == 'human':
            task_content_type = ContentType.objects.get_for_model(instance.workflow)
            for approver in Approver.objects.filter(task_id=instance.name, content_type=task_content_type):
                instance.assignees.add(approver.user)
```

Now I have proper "My Tasks" dashboards with pending/complete tasks.  

The main problem I have faced with joeflow is referencing the `Task` object that is associated with a machine task.

For example, I have not been able to implement the following (simplified) workflow.  The idea is to email someone that they've been assigned and provide a link to the next `UpdateView` but those `Task` objects aren't created yet and I'm unable to reference the `Task` object from within the machine task method... hence the signal code.

```
class ApproveWorkflow(Workflow):

    # fields...

    start = tasks.Start()

    prelim_approve = UpdateView(fields=["approved"])

    def has_approval(self):
        if self.approved:
            return [self.end]

    final_approve = ProtectedUpdateView(fields=["final_approval"])  
    
    def assign_task(self):
        # Cannot assign b/c next Task is not created yet
        pass
    
    def notify_assignee(self):
        # send email to assignee with link to next human step
        pass

    def end(self):
        pass

    edges = (
        (start, assign_task),
        (assign_task, notify_assignee),
        (notify_assignee, approve),
        (approve, has_approval),
        (has_approval, end),
    )

```

**Comment by codingjoe** [comment]
Hi @jordotech,

Thanks for your patience. I needed a little to find time for a thorough response.

No request. Great, then that's how we will build it.

Tasks assignment is a tricky thing, I agree. Especially if you have want to assign a task that does not exist yet.
Let me try to understand better, what you want to achieve. You are looking to build a view, where people can assign a task to someone, right? Would it be sufficient to have a task inbox, where people find their own and unassigned tasks? It seems you also want to notify assignees, after they have been assigned a task right?

I would probably separate the two scenarios. One where tasks get auto-assigned and one where tasks get assigned by humans.

For the latter a simple task create method should be sufficient. For the further we will need two views. One that resembles a tasks inbox and one that handles reassignments and notifications. If a user can reassign a tasks or not could be restricted by permissions. However, those should probably be per process not per task. A task could have a `has_perm` function to decide on a per task base if a user has the permission to reassign a task.

If that worked for you, I would go ahead and try to implement a solution over the weekend and add documentation on how to extend the default behavior.

Best,
Joe

**Comment by codingjoe** [comment]
@jordotech I worked on a draft this weekend on how to do task assignment. I'd really like your feedback. I will try to find time soon to also develop a proper task inbox.

**Comment by jordotech** [comment]
Thanks for that!  I just reviewed the additions, looking good.  

Regarding the [UpdateWithPrevUserView](https://github.com/codingjoe/joeflow/commit/921fdc2047ad5cde033837f80a86cdd87ba5485d#diff-9c7ae79b737b02ffb773e83af9bf7ec3211a1ad50d6f59913f89bec240b1cbf8R4):
That human form not have any inputs, correct?  Would it not be better to present a select input of all available assignees?  The form could initialize with the `prev_task.completed_by_user` (if `prev_task` supplied).  I think it would cover both scenarios.  But yea, this is an important addition to the package, well done.

For reference, this is a flowchart of the real workflow I've developed with joeflow: 
![IT Request Diagram](https://user-images.githubusercontent.com/485972/96466757-c848c400-11ef-11eb-8d02-9b64a58bf334.jpg)

In this scenario there are 2 Django users who are always auto-assigned as "preliminary approver" and "final approver".

**Comment by codingjoe** [comment]
Yes, I too thought of a select input, similar to how issues and pull-requests are assigned here on GitHub. I am finishing up the `create_task` part now and release it. The whole inbox and reassignment part, should go in a separate PR, to keep change sets small.
